client.go

```diff
@@ -214,122 +262,131 @@ func (c *httpClient) doRequest(ctx context.Context, method, url string, params R
 	return err
 }
 
-// nolint:unused
-func (c *httpClient) shouldRetry(resp *http.Response, err error) bool {
+func (c *Client) shouldRetry(resp *http.Response, err error) bool {
 	for _, retryConditional := range c.retryConditionals {
 		if retryConditional(resp, err) {
+			log.Printf("[INFO] Received error %v - Retrying", err)
 			return true
 		}
 	}
 	return false
 }
 
-// nolint:unused
-func (c *httpClient) createRequest(ctx context.Context, method, url string, params RequestParams) (*http.Request, *bytes.Buffer, error) {
+// nolint:nestif
+func (c *Client) createRequest(ctx context.Context, method, endpoint string, params requestParams) (*http.Request, *bytes.Buffer, error) {
 	var bodyReader io.Reader
 	var bodyBuffer *bytes.Buffer
 
 	if params.Body != nil {
-		bodyBuffer = new(bytes.Buffer)
-		if err := json.NewEncoder(bodyBuffer).Encode(params.Body); err != nil {
-			if c.debug && c.logger != nil {
-				c.logger.Errorf("failed to encode body: %v", err)
-			}
-			return nil, nil, fmt.Errorf("failed to encode body: %w", err)
+		reader, ok := params.Body.(*bytes.Reader)
+		if !ok {
+			return nil, nil, c.ErrorAndLogf("failed to read body: params.Body is not a *bytes.Reader")
+		}
+
+		// Reset the body position to the start before using it
+		_, err := reader.Seek(0, io.SeekStart)
+		if err != nil {
+			return nil, nil, c.ErrorAndLogf("failed to seek to the start of the body: %v", err.Error())
 		}
-		bodyReader = bodyBuffer
+
+		bodyReader = reader
 	}
 
-	req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
+	req, err := http.NewRequestWithContext(ctx, method, fmt.Sprintf("%s/%s", strings.TrimRight(c.hostURL, "/"),
+		strings.TrimLeft(endpoint, "/")), bodyReader)
```

Comment:
I think it would be easier to utilize `net/url` package to handle URL concatenation stuff?

Reply:
I tried doing this initially but was running into a lot of issues with the way net/url handled special characters by default.